tests/test_lora_ga.py

```diff
@@ -0,0 +1,284 @@
+                model.zero_grad()
+                outputs = model(inputs)
+                loss = torch.nn.functional.cross_entropy(outputs, labels)
+                loss.backward()
+
+        lora_ga_config = LoraGAConfig(direction="ArB2r", scale="stable")
+        lora_config = LoraConfig(
+            r=4,
+            lora_alpha=8,
+            target_modules=["0", "2"],
+            init_lora_weights="lora_ga",
+            lora_ga_config=lora_ga_config,
+        )
+
+        # Run preprocessing
+        preprocess_loraga(model, lora_config, train_step)
+
+        # Check that gradients were attached
+        assert hasattr(model[0], "_peft_loraga_grad")
+        assert hasattr(model[2], "_peft_loraga_grad")
+        assert model[0]._peft_loraga_grad.shape == model[0].weight.shape
+        assert model[2]._peft_loraga_grad.shape == model[2].weight.shape
+
+    def test_preprocess_without_lora_ga_config_raises(self):
+        model = self.get_dummy_model()
+
+        def train_step():
+            pass
+
+        lora_config = LoraConfig(r=4, lora_alpha=8, target_modules=["0"])
+
+        with pytest.raises(ValueError, match="If you want to use LoRA-GA"):
+            preprocess_loraga(model, lora_config, train_step)
+
+    def test_init_without_lora_ga_config_raises(self):
+        model = self.get_dummy_model()
+        model.train()
+
+        def train_step():
+            for _ in range(4):
+                inputs = torch.randn(2, 10)
+                labels = torch.randint(0, 5, (2,))
+                model.zero_grad()
+                outputs = model(inputs)
+                loss = torch.nn.functional.cross_entropy(outputs, labels)
+                loss.backward()
+
+        # Properly preprocess with lora_ga_config
+        lora_ga_config = LoraGAConfig(direction="ArB2r", scale="stable")
+        lora_config = LoraConfig(
+            r=4,
+            lora_alpha=8,
+            target_modules=["0", "2"],
+            init_lora_weights="lora_ga",
+            lora_ga_config=lora_ga_config,
+        )
+        preprocess_loraga(model, lora_config, train_step)
+
+        # Now try to create a config without lora_ga_config but with init_lora_weights="lora_ga"
+        bad_config = LoraConfig(
+            r=4,
+            lora_alpha=8,
+            target_modules=["0", "2"],
+            init_lora_weights="lora_ga",
+            lora_ga_config=None,  # Missing lora_ga_config!
+        )
+
+        # This should raise an error during get_peft_model
+        with pytest.raises(ValueError, match="lora_ga_config must be provided"):
+            get_peft_model(model, bad_config)
+
+
+@pytest.fixture
+def simple_model():
+    """Fixture providing a fresh simple sequential model for each test."""
+
+    def _make_model():
+        model = torch.nn.Sequential(torch.nn.Linear(10, 10))
+        model.train()
+        return model
+
+    return _make_model
+
+
+@pytest.fixture
+def simple_train_step():
```

Comment:
Similarly, you can return `train_step` directly. Use `simple_model` as an input for this fixture.